File: src/main/java/xyz/deathsgun/modmanager/providers/modrinth/Modrinth.java
/*
 * Copyright 2021 DeathsGun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.deathsgun.modmanager.providers.modrinth;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import net.minecraft.client.MinecraftClient;
import net.minecraft.text.TranslatableText;
import org.apache.http.client.utils.URIBuilder;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import xyz.deathsgun.modmanager.api.mod.Category;
import xyz.deathsgun.modmanager.api.mod.DetailedMod;
import xyz.deathsgun.modmanager.api.mod.ModVersion;
import xyz.deathsgun.modmanager.api.mod.SummarizedMod;
import xyz.deathsgun.modmanager.api.provider.IModProvider;
import xyz.deathsgun.modmanager.api.provider.Sorting;
import xyz.deathsgun.modmanager.providers.modrinth.model.Mod;
import xyz.deathsgun.modmanager.providers.modrinth.model.SearchResponse;
import xyz.deathsgun.modmanager.providers.modrinth.model.Version;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Modrinth implements IModProvider {

    private final Logger logger = LogManager.getLogger("Modrinth");
    private final ArrayList<Category> categories = new ArrayList<>();
    private final HashMap<String, List<SummarizedMod>> cachedCategoryRequests = new HashMap<>();
    private final Gson gson = new Gson();
    private final String baseUrl = "https://api.modrinth.com";
    private final HttpClient http = HttpClient.newHttpClient();

    @Override
    public String getName() {
        return "Modrinth";
    }

    @Override
    public List<Category> getCategories() throws Exception {
        if (!this.categories.isEmpty()) {
            return this.categories;
        }
        logger.debug("Getting categories");
        HttpRequest request = HttpRequest.newBuilder().GET().uri(URI.create(this.baseUrl + "/api/v1/tag/category")).build();
        HttpResponse<String> response = this.http.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            logger.error("Received an invalid status code while getting the categories {}: {}", response.statusCode(), response.body());
            throw new Exception(response.body());
        }
        ArrayList<String> categoryStrings = gson.fromJson(response.body(), new TypeToken<ArrayList<String>>() {
        }.getType());
        categories.clear();
        for (String category : categoryStrings) {
            this.categories.add(new Category(category, new TranslatableText(String.format("modmanager.category.%s", category))));
        }
        return categories;
    }

    @Override
    public List<SummarizedMod> getMods(Sorting sorting, int page, int limit) throws Exception {
        logger.debug("Getting a general list of mods");
        URIBuilder uriBuilder = new URIBuilder(this.baseUrl + "/api/v1/mod");
        uriBuilder.addParameter("index", sorting.name());
        uriBuilder.addParameter("filters", "categories=\"fabric\"");
        return getSummarizedMods(page, limit, uriBuilder);
    }

    @Override
    public List<SummarizedMod> getMods(Category category, int page, int limit) throws Exception {
        logger.debug("Getting category '{}' from Modrinth", category.id());
        String key = String.format("%s|%d|%d", category.id(), page, limit);
        if (this.cachedCategoryRequests.containsKey(key)) {
            return this.cachedCategoryRequests.get(key);
        }
        URIBuilder uriBuilder = new URIBuilder(this.baseUrl + "/api/v1/mod");
        uriBuilder.addParameter("filters", String.format("categories=\"fabric\" AND NOT client_side=\"unsupported\" AND categories=\"%s\"", category.id()));
        List<SummarizedMod> mods = getSummarizedMods(page, limit, uriBuilder);
        this.cachedCategoryRequests.put(key, mods);
        return mods;
    }

    @Override
    public List<SummarizedMod> getMods(@NotNull String query, int page, int limit) throws Exception {
        logger.debug("Searching for '{}' in Modrinth", query);
        URIBuilder uriBuilder = new URIBuilder(this.baseUrl + "/api/v1/mod");
        uriBuilder.addParameter("query", query);
        uriBuilder.addParameter("filters", "categories=\"fabric\" AND NOT client_side=\"unsupported\"");
        return getSummarizedMods(page, limit, uriBuilder);
    }

    @NotNull
    private ArrayList<SummarizedMod> getSummarizedMods(int page, int limit, URIBuilder uriBuilder) throws Exception {
        uriBuilder.addParameter("version", String.format("versions=%s", MinecraftClient.getInstance().getGame().getVersion().getReleaseTarget()));
        uriBuilder.addParameter("offset", String.valueOf(page * limit));
        uriBuilder.addParameter("limit", String.valueOf(limit));
        logger.debug("Using {} for query", uriBuilder.toString());
        HttpRequest request = HttpRequest.newBuilder().GET().uri(uriBuilder.build()).build();
        HttpResponse<String> response = this.http.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new Exception(response.body());
        }
        SearchResponse searchResponse = gson.fromJson(response.body(), SearchResponse.class);
        return searchResponse.toSummarizedMods();
    }

    @Override
    public DetailedMod getMod(@NotNull String id) throws Exception {
        id = id.replaceFirst("local-", "");
        HttpRequest request = HttpRequest.newBuilder().GET().uri(URI.create(this.baseUrl + "/api/v1/mod/" + id)).build();
        HttpResponse<String> response = this.http.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new Exception(response.body());
        }
        return gson.fromJson(response.body(), Mod.class).toDetailedMod();
    }

    @Override
    public List<ModVersion> getVersionsForMod(String id) throws Exception {
        ArrayList<ModVersion> result = new ArrayList<>();
        HttpRequest request = HttpRequest.newBuilder().GET().uri(URI.create(this.baseUrl + "/api/v1/mod/" + id + "/version")).build();
        HttpResponse<String> response = this.http.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new Exception(response.body());
        }
        return Version.toModVersion(gson.fromJson(response.body(), new TypeToken<ArrayList<Version>>() {
        }.getType()));
    }
}

File: src/main/kotlin/xyz/deathsgun/modmanager/providers/modrinth/Modrinth.kt
/*
 * Copyright 2021 DeathsGun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.deathsgun.modmanager.providers.modrinth

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import net.minecraft.text.TranslatableText
import org.apache.http.client.utils.URIBuilder
import org.apache.logging.log4j.LogManager
import xyz.deathsgun.modmanager.ModManager
import xyz.deathsgun.modmanager.api.http.CategoriesResult
import xyz.deathsgun.modmanager.api.http.ModResult
import xyz.deathsgun.modmanager.api.http.ModsResult
import xyz.deathsgun.modmanager.api.http.VersionResult
import xyz.deathsgun.modmanager.api.mod.*
import xyz.deathsgun.modmanager.api.provider.IModProvider
import xyz.deathsgun.modmanager.api.provider.IModUpdateProvider
import xyz.deathsgun.modmanager.api.provider.Sorting
import xyz.deathsgun.modmanager.providers.modrinth.models.DetailedMod
import xyz.deathsgun.modmanager.providers.modrinth.models.ModrinthVersion
import xyz.deathsgun.modmanager.providers.modrinth.models.SearchResult
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

@OptIn(ExperimentalSerializationApi::class)
class Modrinth : IModProvider, IModUpdateProvider {

    private val logger = LogManager.getLogger("Modrinth")
    private val categories: ArrayList<Category> = ArrayList()
    private val cache: HashMap<String, ModsResult> = HashMap()
    private val baseUri = "https://api.modrinth.com"
    private val http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build()


    override fun getName(): String {
        return "Modrinth"
    }

    override fun getCategories(): CategoriesResult {
        if (this.categories.isNotEmpty()) {
            return CategoriesResult.Success(this.categories)
        }
        val request = HttpRequest.newBuilder().GET().setHeader("User-Agent", "ModManager " + ModManager.getVersion())
            .uri(URI.create("${baseUri}/api/v1/tag/category")).build()
        return try {
            val response = this.http.send(request, HttpResponse.BodyHandlers.ofString())
            val categories = json.decodeFromString<List<String>>(response.body())
            for (category in categories) {
                this.categories.add(
                    Category(
                        category,
                        TranslatableText(String.format("modmanager.category.%s", category))
                    )
                )
            }
            CategoriesResult.Success(this.categories)
        } catch (e: Exception) {
            logger.error("Error while getting categories: ", e.message)
            CategoriesResult.Error(TranslatableText("modmanager.error.failedToParse", e.message), e)
        }
    }

    override fun getMods(sorting: Sorting, page: Int, limit: Int): ModsResult {
        val builder = URIBuilder("${baseUri}/api/v1/mod")
        builder.addParameter("index", sorting.name.lowercase())
        builder.addParameter("filters", "categories=\"fabric\"")
        return getMods(builder, page, limit)
    }

    override fun getMods(category: Category, page: Int, limit: Int): ModsResult {
        val key = java.lang.String.format("%s|%d|%d", category.id, page, limit)
        if (this.cache.containsKey(key)) {
            val cachedResult = this.cache[key]!!
            if (cachedResult is ModsResult.Success) {
                return cachedResult
            }
        }
        val builder = URIBuilder("${baseUri}/api/v1/mod")
        builder.addParameter(
            "filters",
            String.format(
                "categories=\"fabric\" AND NOT client_side=\"unsupported\" AND categories=\"%s\"",
                category.id
            )
        )
        return try {
            val result = getMods(builder, page, limit)
            if (result is ModsResult.Success) {
                this.cache[key] = result
            }
            result
        } catch (e: Exception) {
            ModsResult.Error(TranslatableText("modmanager.error.unknown", e), e)
        }
    }

    override fun search(query: String, page: Int, limit: Int): ModsResult {
        val builder = URIBuilder("${baseUri}/api/v1/mod")
        builder.addParameter("query", query)
        builder.addParameter("filters", "categories=\"fabric\" AND NOT client_side=\"unsupported\"")
        return try {
            getMods(builder, page, limit)
        } catch (e: Exception) {
            ModsResult.Error(TranslatableText("modmanager.error.unknown", e.message), e)
        }
    }

    private fun getMods(builder: URIBuilder, page: Int, limit: Int): ModsResult {
        builder.addParameter(
            "version",
            String.format("versions=%s", ModManager.getMinecraftVersion())
        )
        builder.addParameter("offset", (page * limit).toString())
        builder.addParameter("limit", limit.toString())
        val request = HttpRequest.newBuilder().GET().setHeader("User-Agent", "ModManager " + ModManager.getVersion())
            .uri(builder.build()).build()
        val response = this.http.send(request, HttpResponse.BodyHandlers.ofString())
        if (response.statusCode() != 200) {
            return ModsResult.Error(TranslatableText("modmanager.error.invalidStatus", response.statusCode()))
        }
        return try {
            val result = json.decodeFromString<SearchResult>(response.body())
            ModsResult.Success(result.toList())
        } catch (e: Exception) {
            logger.error("Error while requesting mods", e.message)
            ModsResult.Error(TranslatableText("modmanager.error.failedToParse", e.message))
        }
    }

    private val json = Json {
        this.ignoreUnknownKeys = true
    }

    override fun getMod(id: String): ModResult {
        id.replaceFirst("local-", "")
        val request = HttpRequest.newBuilder().GET()
            .setHeader("User-Agent", "ModManager " + ModManager.getVersion())
            .uri(URI.create("${baseUri}/api/v1/mod/${id}")).build()
        val response = try {
            this.http.send(request, HttpResponse.BodyHandlers.ofString())
        } catch (e: Exception) {
            logger.error("Error while getting mod", e.message)
            return ModResult.Error(TranslatableText("modmanager.error.network", e.message), e)
        }
        if (response.statusCode() != 200) {
            return ModResult.Error(TranslatableText("modmanager.error.invalidStatus", response.statusCode()))
        }
        return try {
            val result = json.decodeFromString<DetailedMod>(response.body())
            val categoriesList = ArrayList<Category>()
            result.categories.forEach { categoryId ->
                categoriesList.add(
                    Category(
                        categoryId,
                        TranslatableText("modmanager.category.${categoryId}")
                    )
                )
            }
            ModResult.Success(
                Mod(
                    id = result.id.replaceFirst("local-", ""),
                    slug = result.slug,
                    author = null,
                    name = result.title,
                    shortDescription = result.description,
                    iconUrl = result.iconUrl,
                    description = result.body,
                    license = result.license.name,
                    categories = categoriesList
                )
            )
        } catch (e: Exception) {
            ModResult.Error(TranslatableText("modmanager.error.failedToParse", e.message), e)
        }
    }

    override fun getVersionsForMod(id: String): VersionResult {
        val request = HttpRequest.newBuilder().GET()
            .setHeader("User-Agent", "ModManager " + ModManager.getVersion())
            .uri(URI.create("${baseUri}/api/v1/mod/${id}/version")).build()
        val response = try {
            this.http.send(request, HttpResponse.BodyHandlers.ofString())
        } catch (e: Exception) {
            logger.error("Error while getting mod", e.message)
            return VersionResult.Error(TranslatableText("modmanager.error.network"), e)
        }
        if (response.statusCode() != 200) {
            return VersionResult.Error(TranslatableText("modmanager.error.invalidStatus", response.statusCode()))
        }
        return try {
            val modrinthVersions = json.decodeFromString<List<ModrinthVersion>>(response.body())
            val versions = ArrayList<Version>()
            for (modVersion in modrinthVersions) {
                val assets = ArrayList<Asset>()
                for (file in modVersion.files) {
                    assets.add(Asset(file.url, file.filename, file.hashes, file.primary))
                }
                versions.add(
                    Version(
                        modVersion.version,
                        modVersion.changelog,
                        getVersionType(modVersion.type),
                        modVersion.gameVersions,
                        assets
                    )
                )
            }
            VersionResult.Success(versions)
        } catch (e: Exception) {
            VersionResult.Error(TranslatableText("modmanager.error.failedToParse", e.message), e)
        }
    }

    private fun getVersionType(id: String): VersionType {
        return when (id) {
            "release" -> VersionType.RELEASE
            "alpha" -> VersionType.ALPHA
            "beta" -> VersionType.BETA
            else -> VersionType.UNKNOWN
        }
    }

}